driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java

```diff
@@ -136,4 +136,20 @@ public static <R extends Expression> R ofNull() {
         return new MqlExpression<>((cr) -> new AstPlaceholder(new BsonNull()))
                 .assertImplementsAllExpressions();
     }
+
+    static NumberExpression numberToExpression(final Number number) {
+        if (number instanceof Integer) {
+            return of((int) number);
+        } else if (number instanceof Long) {
+            return of((long) number);
+        } else if (number instanceof Double) {
+            return of((double) number);
+        } else if (number instanceof Decimal128) {
+            return of((Decimal128) number);
+        } else if (number instanceof BigDecimal) {
+            return of((BigDecimal) number);
+        } else {
+            throw new IllegalArgumentException("Number must be an Integer, Long, Double, Decimal128, or BigDecimal");
+        }
```

Comment:
I disagree that the API is intended to use codecs. There is a codec that is supposed to convert expressions to BSON, so the codecs depend on this exp API existing, but the exp API does not depend on any codecs.

We are also leaving open the possibility of moving to an AST, the point of which is to decouple expressions from the BSON/MQL, and this would seem to go in the opposite direction.

Reply:
It's not a public method, so the thing with using codecs is as irrelevant as the thing with renaming. However, it is still useful to talk about the usage of `CodecRegistry` in `MqlExpression`.

> There is a codec that is supposed to convert expressions to BSON, so the codecs depend on this exp API existing, but the exp API does not depend on any codecs.

Currently, `MqlExpression.toBsonValue` (which is how `MqlExpressionCodec` encodes `MqlExpression` to BSON) passes `CodecRegistry` to `MqlExpression.fn`, yet as far as I understand, there are no places where this `CodecRegistry` is used. What is the planned use for the `CodecRegistry` that is passed to `MqlExpression.fn`?

Reply:
I added a comment there some time ago: https://github.com/mongodb/mongo-java-driver/pull/1032#discussion_r1013283977. Thanks for the `DocumentExpression`!